### ISSUE TITLE
Geocoding function to transform addresses to a lat-lng using Open Street Map

### ISSUE THREAD

**Issue by sunnymui** [codeforsanjose/disaster-response-sj#55]
Research how to geocode an address text string into a lat/lng with open street map, work on a function to implement that in the app.

**Comment by sunnymui** [comment]
Some starting research.

[Nominatim is the Open Street Maps geocoder tool](https://wiki.openstreetmap.org/wiki/Nominatim) to search OSM data with a query string or structured address parameters and return place data, including the latitude and longitude. [Demo here.](https://nominatim.openstreetmap.org/search)
- Downside to the OSM Foundation's Nominatim API is no heavy use (1 request per second max) and no autocomplete through the API on the client side. Though you could always host your own instance of Nominatim to do whatever you want. [Nominatim API Usage Policy here](https://operations.osmfoundation.org/policies/nominatim/). 
- Might be some interesting issues to look into, [read more here, though this article is a year old](https://blog.emacsen.net/blog/2018/02/16/osm-is-in-trouble/)
- There are some 3rd party/alternative hosted geocoder API services using OSM that support heavier usage:
  1.  [Mapquest Open Search Nominatim API](https://developer.mapquest.com/documentation/open/nominatim-search/) - 15,000 requests/mon for free tier, no published rate limits, at least [according to this thread](https://developer.mapquest.com/forum/rate-limits-open-endpoint).
  2. [Open Cage Data](https://opencagedata.com/pricing) - ~75,000 requests/mon (2,500/day), max 1 request/sec for free tier.
  3. [Location IQ](https://locationiq.com/pricing) - ~300,000 requests/mon (10k/day), max 2 requests/sec, 60 req/min for free tier

There's also some open source geocoder alternatives to Nominatim you can host your own instances of, though it mos def looks like a big undertaking with a lot of moving parts:
  1. [Photon](https://github.com/komoot/photon) - Built with ElasticSearch and apparently typo tolerant and multilingual. To host your own you have to download the 53 gb compressed search index. They actually have a demo with an autocomplete feature on their main site. They have a [public API](https://photon.komoot.de/) though it doesn't seem to be intended for extensive usage. They seem to have a [prewritten autocomplete module](https://github.com/komoot/typeahead-address-photon) if you don't wanna write your own. 
  2. [Pelias](https://github.com/pelias/pelias) - Built with Node.js and ElasticSearch. Also a part of the Linux Foundation now. Supports autocomplete. Looks like a pretty modular design and [fairly good documentation](https://github.com/pelias/documentation). You can setup a quick Docker instance for a small city sized area quickly for a baby project. For more serious installations for production, they recommend Kubernetes--[install info here](https://github.com/pelias/documentation/blob/master/getting_started_install.md). For people that don't want to deal with any of the management of a geocoder setup, there is a [cloud based enterprise grade API offered by Geocode Earth](https://geocode.earth/cloud), starting at $200/mon for 200k requests/mon @ 10 requests/sec.

Anyway, TLDR can just use one of the hosted API's for now to get stuff running. Autocomplete may take some work.

**Comment by sunnymui** [comment]
Thinking a bit more into the integration side of this:

Right now on the admin panel page you enter an address and the validations are all client side during form filling against specified rules in the js. 

The UX/UI for geocoding that address could be something like where you enter the address, then click submit (maybe change the label here). It makes an ajax call to the open street maps api of choice to geocode that address into a lat-lng pair. Returns that lat lng and plots it on a map instance so the user can verify the location is correct visually. The submit button is changed into a Confirm Location button. The user then clicks the confirm location button to actually send that data to the server.

Additional function is the autocomplete integration, which would allow user to select an address once they start typing it into the form fields and have it populate all the fields with for sure valid data. A map instance is shown with the location plotted for visual verification. We can just go straight to a confirm location button here.

**Comment by sunnymui** [comment]
Note for the future: OSM library we're using to plot map tiles is Leaflet.

Thinking about the UX some more:

The user may have coordinates for a disaster but not an address if it's in a remote area. Perhaps let users enter a disaster by coordinates OR by address. That would mean only requiring either address form or coordinates form be entered.

Enhancement idea: Geocoded coordinates should be plotted on a confirmation map tile, but maybe be draggable to allow adjusting the location in an easy way?

Regarding latitude-longitude input:

[Good discussion on the ux of lat long form inputs](https://ux.stackexchange.com/questions/16878/standard-for-entering-latitude-longitude-in-text-fields)
Decimal seems like the most user accessible format and the one most commonly used. DMS format is also used and an integrated converter option to use that format may also be a warranted feature enhancement.

[More on converting DMS to Decimal Lat Long here.](https://www.latlong.net/degrees-minutes-seconds-to-decimal-degrees)

**Comment by sunnymui** [comment]
Resolved in PR https://github.com/codeforsanjose/disaster-response-sj/pull/72/
